problem_builder/table.py

```diff
@@ -66,23 +74,122 @@ class MentoringTableBlock(StudioEditableXBlockMixin, StudioContainerXBlockMixin,
+    view_options = Dict(scope=Scope.user_state_summary)
     has_children = True
 
-    def student_view(self, context):
-        context = context.copy() if context else {}
-        fragment = Fragment()
+    css_path = 'public/css/mentoring-table.css'
+    js_path = 'public/js/review_blocks.js'
+
+    @XBlock.json_handler
+    def table_render(self, data, suffix=''):
+        context = {}
         header_values = []
         content_values = []
+        user_service = self.runtime.service(self, 'user')
+        target_username = data.get('target_username')
+        if user_service and target_username:
+            username = user_service.get_current_user().opt_attrs['edx-platform.username'].lower()
+            permitted = self.view_options.get(username)
+            if target_username in permitted:
+                context['student_submissions_key'] = user_service.get_anonymous_user_id(
+                    target_username, unicode(self.runtime.course_id)
+                )
+            elif target_username == username:
+                # Use default settings.
+                pass
+            else:
+                raise JsonHandlerError(403, _("You are not permitted to view this student's table."))
+
         for child_id in self.children:
             child = self.runtime.get_block(child_id)
             # Child should be an instance of MentoringTableColumn
             header_values.append(child.header)
             child_frag = child.render('mentoring_view', context)
             content_values.append(child_frag.content)
-            fragment.add_frag_resources(child_frag)
         context['header_values'] = header_values if any(header_values) else None
         context['content_values'] = content_values
+        html = loader.render_template('templates/html/mentoring-table.html', context)
+        return {'content': html}
+
+    @property
+    def current_user_key(self):
+        user = self.runtime.service(self, 'user').get_current_user()
+        # We may be in the SDK, in which case the username may not really be available.
+        return user.opt_attrs.get('edx-platform.username', 'username').lower()
+
+    @XBlock.json_handler
+    def get_shared_list(self, data, suffix=''):
+        context = {'shared_with': self.shared_with.get(self.current_user_key, [])}
+        return {
+            'content': loader.render_template('templates/html/mentoring-table-shared-list.html', context)
+        }
+
+    @XBlock.json_handler
+    def share_results(self, data, suffix=''):
+        target_usernames = data.get('usernames')
+        target_usernames = [username.strip().lower() for username in target_usernames if username.strip()]
+        current_username = self.current_user_key
+        if not target_usernames:
+            raise JsonHandlerError(400, _('Usernames not provided.'))
+        shared_with = self.shared_with.get(current_username, [])
```

Comment:
@Kelketek shared_with and view_options could be sets - in such case it wouldn't be necessary to check if it's already there - just add to a set.